File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # [-4, -1, -1, 0, 1, 2]
        so = sorted(nums)
        res = set()
        for k in range(len(so) - 2):
            # target = 1, i = 2, j = 5
            target = -so[k]
            i, j = k + 1, len(so) - 1
            while i < j:
                s = so[i] + so[j]
                if s < target:
                    i += 1
                    continue
                if s > target:
                    j -= 1
                    continue
                res.add((so[k], so[i], so[j]))
                i += 1
        return list(res)

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        dfs(root, res, 0);
        return res;
    }
private:
    void dfs(TreeNode* node, vector<int>& nums, int depth) {
        if (!node) {
            return;
        }
        if (nums.size() == depth) {
            nums.push_back(node->val);
        }
        nums[depth] = node->val;
        dfs(node->left, nums, depth + 1);
        dfs(node->right, nums, depth + 1);
    }
};
